src/inmanta/plugins.py

```diff
@@ -201,16 +216,99 @@ def type_string(self) -> str:
     def type_string_internal(self) -> str:
         return self.type_string()
 
+    def as_python_type_string(self) -> "str | None":
+        return "None"
+
+    def corresponds_to(self, pytype: typing.Type) -> bool:
+        return pytype is type(None)
+
+    def has_custom_to_python(self) -> bool:
+        return True
+
+    def to_python(self, instance: object) -> "object":
+        assert instance is None
+        return None
+
 
 # Define some types which are used in the context of plugins.
 PLUGIN_TYPES = {
-    "any": inmanta_type.Type(),  # Any value will pass validation
-    "expression": inmanta_type.Type(),  # Any value will pass validation
+    "any": inmanta_type.AnyType(),  # Any value will pass validation
+    "expression": inmanta_type.AnyType(),  # Any value will pass validation
     "null": Null(),  # Only NoneValue will pass validation
     None: Null(),  # Only NoneValue will pass validation
 }
 
 
+def validate_and_convert_to_python_domain(expected_type: inmanta_type.Type, value: object) -> object:
+    """
+    Given a model domain value and an inmanta type, produce the corresponding python object
+
+    Unknowns are not handled by this method!
+    """
+    import inmanta.ast.entity
+
+    expected_type.validate(value)
+
+    if isinstance(value, NoneValue):
+        # if the type is not nullable, it will fail when validating
+        # if the value is None, it becomes None
+        return None
+
+    base_type = expected_type.get_base_type()
+    if base_type.has_custom_to_python():
+        return expected_type.to_python(value)
```

Comment:
This is probably too strict and will break existing code that is not typed correctly

Reply:
It was a stupid mistake elsewhere I think

Reply:
https://github.com/inmanta/inmanta-core/pull/8399/commits/e26b82d7780e73edc6add0e33d316b7bd5738da6